File: golang/bfs.go
package main

type Node struct {
		Name string
		Children []*Node
}

func (n *Node) BFS(nodes []string) []string {
		queue := []*Node{n}
		for len(queue) > 0 {
				// Popping the node off the queue
				current := queue[0]
				queue = queue[1:]
				nodes = append(nodes, current.Name)
				for _, child := range current.Children {
						queue = append(queue, child)
				}
		}
		return nodes
}

File: java/BFS.java
import java.util.*;

public class BFS {
    static class Node {
        String name;
        List<Node> children = new ArrayList<Node>();

        public Node(String name) {
            this.name = name;
        }

        public List<String> breadthFirstSearch(List<String> arr) {
            Queue<Node> queue = new LinkedList<>();
            queue.add(this);
            while (!queue.isEmpty()) {
                Node current = queue.poll();
                arr.add(current.name);
                queue.addAll(current.children);
            }
            return arr;
        }

        public Node addChild(String name) {
            Node child = new Node(name);
            children.add(child);
            return this;
        }
    }
}
